hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -28,45 +48,102 @@ object InferPType {
+
+        while(it.hasNext) {
+          val irElem = it.next()
+
+          InferPType(irElem, env)
+          assert(head.pType2 isOfType irElem.pType2)
+
+          if(allRequired == true && irElem.pType2.required == false) {
+            allRequired = false
+          }
+        }
+
+        PArray(head.pType2, allRequired)
+      }
+      case MakeStream(irs, _) => {
+        val allRequired = irs.forall(ir => {
+          InferPType(ir, env)
+
+          ir.pType2.required
+        })
+
+        irs(0).pType2.setRequired(allRequired)
+      }
       case MakeNDArray(data, shape, _) => {
         InferPType(data, env)
         InferPType(shape, env)
 
         val nElem = shape.pType2.asInstanceOf[PTuple].size
 
-        PNDArray(coerce[PArray](data.pType2).elementType, nElem)
+        PNDArray(coerce[PArray](data.pType2).elementType, nElem, data.pType2.required)
+      }
+      case ArrayRange(start: IR, stop: IR, step: IR) => {
+        InferPType(start, env)
+        InferPType(stop, env)
+        InferPType(step, env)
+
+        assert(start.pType2 isOfType stop.pType2)
+        assert(start.pType2 isOfType step.pType2)
+
+        val allRequired = start.pType2.required &&
+          stop.pType2.required &&
+          step.pType2.required
+
+        PArray(start.pType2, allRequired)
+      }
+      case StreamRange(start: IR, stop: IR, step: IR) => {
+        InferPType(start, env)
+        InferPType(stop, env)
+        InferPType(step, env)
+
+        assert(start.pType2 isOfType stop.pType2)
+        assert(start.pType2 isOfType step.pType2)
+
+        val allRequired = start.pType2.required &&
+          stop.pType2.required &&
+          step.pType2.required
+
+        PArray(start.pType2, allRequired)
+      }
+      case ArrayLen(a: IR) => {
+        InferPType(a, env)
+
+        a.pType2
+      }
+      // TODO: Infer orderedCollection/first arg?
+      case LowerBoundOnOrderedCollection(_, bound: IR, _) => {
+        InferPType(bound, env)
+        bound.pType2
       }
-      case _: ArrayLen => PInt32()
-      case _: ArrayRange => PArray(PInt32())
-      case _: StreamRange => PStream(PInt32())
-      case _: LowerBoundOnOrderedCollection => PInt32()
       case _: ArrayFor => PVoid
       case _: InitOp => PVoid
       case _: SeqOp => PVoid
       case _: Begin => PVoid
-      case Die(_, t) => PType.canonical(t)
+      case Die(_, t) => PType.canonical(t, true)
       case If(cond, cnsq, altr) => {
         InferPType(cond, env)
         InferPType(cnsq, env)
         InferPType(altr, env)
 
-        assert(cond.pType2 == PBoolean())
+        assert((cnsq.pType2 isOfType altr.pType2) && (cond.pType2 isOfType PBoolean()))
 
-        if (cnsq.pType2 != altr.pType2)
-          cnsq.pType2.deepOptional()
-        else {
-          assert(cnsq.pType2 isOfType altr.pType2)
-          cnsq.pType2
-        }
+        cnsq.pType2.setRequired(cnsq.pType2.required && altr.pType2.required)
```

Comment:
If the condition is optional, then the result is optional as well. (missing condition => missing result)

Reply:
Mentioned in more detail below, but similar to enforcing that NDArrray shapes to be non-missing, shouldn't the condition (or operation), always be non-missing if the If node is non-missing? Meaning If node doesn't depend on the missing ness of its condition, just the missing ness of the operands?

Reply:
> shouldn't the condition (or operation), always be non-missing if the If node is non-missing

What does "If node is non-missing" mean?

Reply:
Here I mean that the condition is always required if the `If` node is defined. How can the condition ever be optional? It doesn't depend on anything else, it is a fact of the If node. Trivially we could have written the If node as `If_true` and `If_false` and omitted the condition altogether.

Reply:
`If(NA(TBoolean), I32(1), I32(2))` is a totally valid node.

Reply:
the condition is an IR, so its requiredness is computed the way all other IRs are. If the condition is missing, the result of the If node is missing.

Reply:
That feels odd. What does if(null) 1 else 2 mean, except that If(false) 1 else 2

Reply:
`If(NA(TBoolean), I32(1), I32(2))` should be simplified to `NA(TInt32)`

Reply:
I think the idea is that we generally treat missingness as R does---it's propagated pretty extensively. The NA propagation does make sense in certain cases:
```
if (field % 2 == 0) "even" else "odd"
```
then the correct behavior is to return `NA` and not `"odd"` if `field` is missing. We introduce logic to allow users to say `NA -> false` for the condition in Python.

Reply:
Consider this use case:
```
array = # some array expression
array.map(lambda x: hl.cond(x < 0, -1, 1))
```
What should this do if `array = [10, -5, NA(TInt32)]`? We have no choice but to evaluate `If(NA < 0, 0, x)`, so we have to return something. I think the only consistent answer is to return `NA`, giving `[1, -1, NA]`. If we treat missing the same as false, this would be `[1, -1, -1]` which would be very surprising.

Reply:
That makes sense!